### ISSUE TITLE
Need least squares regression?

### ISSUE THREAD

**Issue by korintje** [veusz/veusz#579]
I like Veusz fitting for the flexible expression, but it seems sometimes give strange results even in some simple linear regression (as below). Are these intended behaviors or bugs?

If this behavior is correct, I think it is worth separately introducing basic polynomial fitting using the least squares method.
I believe it is not hard to implement using `numpy.polynomial.polynomial.polyfit`.

I am far from an expert of fitting algorithm, so I would like to ask opinions from the author, other contributors, and users.

![bar](https://user-images.githubusercontent.com/30950088/150340165-045b547f-c206-47e3-a0c6-7935d31e92cf.png)

**Comment by T0T0R** [comment]
TLDR : I also had several issues regarding linear regressions, but they are usually fixed by adding error values to your dataset.
According to the author, this functionnality still needs improvements ([see manual](https://veusz.github.io/docs/manual/introduction.html#widget)).

The fit in Veusz is done by minimizing a quantity `c` ([source there](https://github.com/veusz/veusz/blob/5a1e2af5f24df0eb2a2842be51f2997c4999c7fb/veusz/widgets/fit.py#L61) or `chi2` [here](https://github.com/veusz/veusz/blob/5a1e2af5f24df0eb2a2842be51f2997c4999c7fb/veusz/utils/fitlm.py#L60) if **iminuit** is not available) which is a function of the y-errors. This quantity is the **phi** quantity of Peter Scott's *Intermediate Laboratory and Advanced Laboratory courses* [see chap. 3, eq. 3.4](https://scott.physics.ucsc.edu/pdf/133_draftman.pdf).
When errors are not provided in the dataset, Veusz assumes a [5% error](https://github.com/veusz/veusz/blob/5a1e2af5f24df0eb2a2842be51f2997c4999c7fb/veusz/widgets/fit.py#L280) for each y-value, then does a least-squares fit based on minimising **c**.
In the case of `numpy.polynomial.polynomial.polyfit`, when no errors are provided ([weights are `None`](https://numpy.org/doc/stable/reference/generated/numpy.polynomial.polynomial.polyfit.html)) the fit is done nonetheless by using [weights of 1](https://github.com/numpy/numpy/blob/4adc87dff15a247e417d50f10cc4def8e1c17a03/numpy/polynomial/polyutils.py#L637).

I made a few graphs showing the fit with the Veusz's widget and polyfit function of numpy. In the top-left one, when polyfit have *no* weights the fit is not correct, whereas when weights of 1/(5% of y data) are given, both fits are on top of each other.
In the next 3 graphs, polyfit's weights are either calculated from the error provided in the dataset, or generated with a list in python. You can see that - for different errors - the fitting tool of Veusz is matching polyfit.
![answer](https://user-images.githubusercontent.com/26389071/154811177-fa3e7f16-db9d-4b81-b843-c815d05a79b2.png)
[Here is the code for generating the image, just rename it to ".vsz"](https://github.com/veusz/veusz/files/8102646/fit_example.txt)

**Comment by korintje** [comment]
@T0T0R 
Thank you for the reply and very comprehensive explanations.
I understood the difference between veusz and numpy polyfit in their default regression weight parameter. 

Which behavior do you think is generally more "appropriate"?
In my (naive) opinion, the polyfit behavior of numpy (weight=1 by default) seems more intuitive.

**Comment by T0T0R** [comment]
I agree with you, weights of 1 seem more intuitive to me.
When I want to do a "quick fit" (for instance, to estimate a parameter given by the slope of a lin. regression), I put data in Veusz without specifying errors, and I expect a "simple" fit with weights of 1.

_But_, I think the default behavior of Veusz is actually adequate, because it forces the user to think about the error values : 'If you don't provide a thorough dataset with errors, I will give you an incorrect fit'. I don't think a "quick fit" can exist (I have to say I have no degree in statistics), and I believe that 'Just give me the slope already' is a dangerous approach. This is why **I think the behavior of Veusz is appropriate in applying its own weights: if the user is willing to have a reliable result, he has to thoroughly think about his measurements and the errors.**

Now, regarding the choice of default weights, apparently the author choosed a **relative** error because [he fitted mostly logarithmic data](https://github.com/veusz/veusz/pull/439#issuecomment-715911070). For me (doing chemistry stuff), absolute **errors** on variables are more common and I think I would prefer the default behavior to be with absolute errors.

**Comment by kaugummiheld** [comment]
Thanks @T0T0R for the explanation above. I've struggled quite a bit before I got to the idea that is has something to do with the error values of my datasets when I get "weird" fits. Also, I second that this forces the users to think more about what they're doing when they perform a fit. 

My preference would be to apply an absolute error, since logarithmic weighted fitting of datasets one has to to some preparations anyway, i.e. creating a new dataset `B = np.log(A)` instead of simply changing an axis to logarithmic and call it a day. 
I curious what @jeremysanders thinks of this, roughly 1.5 years later. :)
